### ISSUE TITLE
RFE: add command line option to rpmbuild that exit with error when SPEC file has unexpanded macro

### ISSUE THREAD

**Issue by xsuchy** [rpm-software-management/rpm#2488]
It is pretty common that package maintainers do an mistake in the SPEC file and forget a BuildRequires. It passes for them in plain rpmbuild on their workstation will full set of packages. But when the build is run in Koji or Mock the package that defines macro is missing.  E.g. %py3_build. For obvious reason packagers do not use ${?py3_build}. So the section (or even worse scriptlet) tryies to run the unexpanded string.  Which ends up with cryptic error:

```
+ %py3_build
/var/tmp/rpm-tmp.XYSHTs: line 44: fg: no job control
```

I believe we can avoid this by implement a new command line option for rpmbuild. E.g. `--unexpanded-macros-are-fatal`  and if used, rpmbuild will exit with an error when some macro is not expanded. Undefined macros %{?foo} will be ignored and expanded to nothing as of now.

Mock (and therefore Koji) can use this option and provide better error message to users than the current cryptic ones. 

3rd parties and ordinary users will not be affected because it will be opt-in and without this parameter, the behavior will be unchanged. 

This is the reference that triggered this. But this is not the only one. I see it often.
https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/ZFMKBPSDLGOLM6JYTZLNV2TOEP3ZL6F6/

**Comment by voxik** [comment]
How about using generator to have auto BRs for such macros instead? I think that could be doable these days.

**Comment by pmatilai** [comment]
Right, I totally forgot about we're adding macro provides nowadays.

So in principle that should be doable, the devil is probably in the implementation as an unknown macro is something that happens deep inside the macro engine and the spec parser and there's no means to hook that up into anything at all.

**Comment by xsuchy** [comment]
> How about using generator to have auto BRs for such macros instead?

We can do that for common macros in Fedora (e.g. py3_*) .  But independently. Because this will not solve macros defined e.g. in ~/.rpmmacros or typos.

**Comment by pmatilai** [comment]
Catching unexpanded macros from the preamble would be one thing, but doing that for scriptlet sections is an entirely different matter, because the content in those sections is shell script rather than something rpm understands. So we don't know whether a % in a scriptlet is a legit literal % in a shell script content or an unexpanded macro, because ... rpm has always let the unexpanded macros through as-is. Fixing *that* seems like a pre-requisite for this, and it's something that would need to be eased in gradually, along with some clear-cut rules such as non-macro % needs to be escaped everywhere in the spec.
